### PR TITLE
accounts-db: silence false positive info in shrink

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

We want to know if we find we have zero alive bytes, but all accounts are not zero lamports when shrinking,
which would inidate a bug somewhere in clean. 
Right now this log line is going off when a slot ONLY contains single ref zero lamport accounts. In that case, we need clean to mark the slot dead.

#### Summary of Changes

log when we find NOT all accounts are zero lamports when shrinking.

Fixes #

### CODE REVIEW

**Review by jeffwashington — DISMISSED** [review]
lgtm

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -4347,6 +4347,16 @@ impl AccountsDb {
...
             );
 
+            if !shrink_collect.all_are_zero_lamports {
+                // if all are zero lamports, then we expect that we would like to mark the whole slot dead, but we cannot. That's clean's job.
+                info!(
+                    "Unexpected shrink for slot {} alive {} capacity {}, \
+                    likely caused by a bug for calculating alive bytes. All alive bytes are zero: {}, {}",
+                    slot, shrink_collect.alive_total_bytes, shrink_collect.capacity,
+                    store.alive_bytes(), shrink_collect.zero_lamport_single_ref_pubkeys.len() * aligned_stored_size(0),
+                );
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: There's a lot of duplication with the info!() line directly preceding. Is the intent to log both of these when we are in the not-all-are-zero-lamports case?

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: Fixed. 
Merged two messages into one.

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -4341,11 +4341,21 @@ impl AccountsDb {
...
             );
+            if !shrink_collect.all_are_zero_lamports {
+                // if all are zero lamports, then we expect that we would like to mark the whole slot dead, but we cannot. That's clean's job.
+                msg.push_str(&format!(
+                    "All alive bytes are zero: {}, {}",
+                    store.alive_bytes(),
+                    shrink_collect.zero_lamport_single_ref_pubkeys.len() * aligned_stored_size(0)
+                ));
+            }
+            info!("{}", &msg);
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: I don't want any msg if `all_are_zero_lamports`. This is a false positive and expected. We have zero 'alive' bytes, but we have many alive zero lamport single ref accounts. This is normal behavior and will be resolved by `add_uncleaned_roots` above. I think the whole pr here is `if !shrink_collect.all_are_zero_lamports {` around the existing `info`.

File: accounts-db/src/accounts_db.rs [hunk 3]
```diff
@@ -4347,6 +4347,16 @@ impl AccountsDb {
                 slot, shrink_collect.alive_total_bytes, shrink_collect.capacity
             );
 
+            if !shrink_collect.all_are_zero_lamports {
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs:4345]: ah, answer is to remove the `info` above on line 4345. This is the one that goes in the `if`. Sorry - I misread the original pr.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:4345]: updated.

**Review by jeffwashington — DISMISSED** [review]
lgtm!

File: accounts-db/src/accounts_db.rs [hunk 4]
```diff
@@ -4341,11 +4341,20 @@ impl AccountsDb {
...
-                likely caused by a bug for calculating alive bytes.",
-                slot, shrink_collect.alive_total_bytes, shrink_collect.capacity
-            );
+
+            if !shrink_collect.all_are_zero_lamports {
+                // if all are zero lamports, then we expect that we would like to mark the whole slot dead, but we cannot. That's clean's job.
+                info!(
+                    "Unexpected shrink for slot {} alive {} capacity {}, \
+                        likely caused by a bug for calculating alive bytes. \
+                        All alive bytes are zero: {}, {}",
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db.rs]: For me, I find this confusing. The text says that all the alive bytes are zero, and then we log the alive bytes in this storage. If it is always zero, why log it. If it is non-zero, it seems to contradict the text.

And then there's a second number that's not described. I believe it's the number of *total* stored bytes from zero lamport accounts.

**Review by brooksprumo — COMMENTED** [review]
some nits. Also imo, the PR title feels contradictory to what's happening.

File: accounts-db/src/accounts_db.rs (hunk 4 above)
Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: `Self::should_not_shrink` is another reason we might get here.

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: we can get rid of the last 2 args.

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: backing up, since shrink treats single ref zero lamport accounts as dead, we can now correctly hit `num_alive_bytes`=0. This logs and says we have a bug. But, we don't. So, we need to silence that misleading info.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: Updated.

File: accounts-db/src/accounts_db.rs [hunk 5]
```diff
@@ -4341,11 +4341,15 @@ impl AccountsDb {
...
-                "Unexpected shrink for slot {} alive {} capacity {}, \
-                likely caused by a bug for calculating alive bytes.",
-                slot, shrink_collect.alive_total_bytes, shrink_collect.capacity
-            );
+
+            if !shrink_collect.all_are_zero_lamports {
+                // if all are zero lamports, then we expect that we would like to mark the whole slot dead, but we cannot. That's clean's job.
+                info!(
+                    "Unexpected shrink for slot {} alive {} capacity {}, \
+                        All alive bytes are zero.",
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: I think revert this change. We are in the `!all_are_zero_lamports` case ;-)

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: ok. reverted!

File: accounts-db/src/accounts_db.rs [hunk 6]
```diff
@@ -4346,8 +4346,13 @@ impl AccountsDb {
...
                     "Unexpected shrink for slot {} alive {} capacity {}, \
-                        All alive bytes are zero.",
-                    slot, shrink_collect.alive_total_bytes, shrink_collect.capacity
+                        likely caused by a bug for calculating alive bytes. \
+                        All alive bytes are zero: {}, {}",
+                    slot,
+                    shrink_collect.alive_total_bytes,
+                    shrink_collect.capacity,
+                    store.alive_bytes(),
+                    shrink_collect.zero_lamport_single_ref_pubkeys.len() * aligned_stored_size(0)
```
Comment (jeffwashington) [inline accounts-db/src/accounts_db.rs]: we don't want these lines back...
shrink_collect.zero_lamport_single_ref_pubkeys.len() * aligned_stored_size(0)

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: Here is what I understand.

When we call `do_shrink_slot_store`, we check that if the `should_not_shrink = true or alive_total_bytes = 0, we skip and don't shrink. 

If we skip shrinking, we want to log some info, when we think the skipping is an unexpected case. 

In the old code, we log for every skipped store regardless. However, for case when all accounts in the store are zero_lamports and we skip, which should BE expected. Therefore, we shouldn't log it. 

That's what the new change in PR is about - only log when we skip shrink and the account are NOT all zero lamports.

Is this understanding correct?

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs]: yes.

So, revert the text change so that it remains: `likely caused by a bug for calculating alive bytes.",`

**Review by brooksprumo — APPROVED** [review]
:shipit:

File: accounts-db/src/accounts_db.rs (hunk 6 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: ok. reverted to use original log text.

**Review by jeffwashington — APPROVED** [review]
lgtm
